notes/dns-client-proxy.md

```diff
@@ -0,0 +1,535 @@
+
+DoT
+
+: DNS over TLS as described in [@RFC7858]
+
+DoQ
+
+: DNS over QUIC ([@RFC9000]) as described in [I-D.ietf-dprive-dnsoquic],
+not to be confused with DNS over HTTP/3 which also uses QUIC
+
+EDNS(0) Option
+
+: An option as described in [@RFC6891]
+
+h2
+
+: This TLS ALPN identifies HTTP/2 as described in [@RFC7540]
+
+h3
+
+: This TLS ALPN identifies HTTP/3, which is HTTP over QUIC and
+is described in I.D.ietf-quic-http (expired draft)
+
+Interface Name
+
+: A name that identifies a network interface as described in [@RFC3493].
+In addition, an interface index converted to a decimal number is also
+consider an interface name.
+
+# Introduction
+
+The introduction of many new transport protocols for DNS in recent years
+(DoT, DoH, DoQ)
+significantly increases the complexity of DNS stub
+resolvers that want to support these protocols. In addition,
+for short-lived applications, the overhead of setting a DoH connection
+is quite high if the application only needs to send a few DNS requests.
+
+A practical way forward
+is to have a DNS client proxy in the host operating system.
+A local proxy may provide some benefit to short-lived applications by
+caching results. In particular if the system uses a so called 'public
+DNS resolver'. In general we assume that the cache is tagged according
+to the source of a reply and the transport it is received on.
+
+
+This allows
+applications to communicate using Do53 and still get the privacy
+benefits from using more secure protocols over the internet. However,
+such a setup leaves the application with no control over which transport
+the proxy uses. This document introduces EDNS(0) options that allow a
+stub resolver to request certain transports and allow the proxy to report
+capabilities and actual transports that are available.
+
+
+With respect to DNSSEC, we assume that an application that needs
+DNSSEC validation, for example, for DANE validation or SSHFP, will
+perform the DNSSEC validation within the application itself and does not
+trust the proxy. The proxy can of course do DNSSEC validation as well.
+Important however, is that an untrusted proxy cannot provide an application
+with a traditional (unsigned) trust anchor.
+
+For the transport configuration we expect three levels of details. The
+first is a choice between anythings goes, unauthenticated encryption and
+authenticated encryption. The second level is where the application also
+specifies the names and/or IP addresses of upstream resolvers. The
+third level is where the application also specifies which transports
+(Do53, DoT, DoH, DoQ) are allowed to be used. A final transport parameter
+is the outgoing interface that is to be used.
+
+For authentication we can have a mix of PKI and DANE. Options are one of
+the two and not the other, both or one of the two.
+
+A poorly constructed DNS client proxy may forward DNS packets without
+properly handling of EDNS(0) options. To detect and prevent this,
```

Comment:
Yes, I agree that that would be poorly constructed, however there is a RFC for DNS proxies which has ambivalent statements about this when trying to take into account the transparency principle for DNS Proxies. Have a read of section 3 of that RFC: 

https://www.rfc-editor.org/rfc/rfc5625.html#section-3

I guess it's a balance. Perhaps we should replace "poorly constructed" with something a bit less judgmental?